snapcraft/internal/meta/slots.py

```diff
@@ -0,0 +1,195 @@
+
+    def __init__(
+        self,
+        *,
+        slot_name: str = "",
+        read: List[str] = None,
+        write: List[str] = None,
+        use_source_key: bool = True
+    ) -> None:
+        super().__init__(slot_name=slot_name)
+
+        self._plug_dict = dict()  # type: Dict[str, Any]
+        self.interface = "content"  # type: str
+
+        if read is None:
+            self.read = list()  # type: List[str]
+        else:
+            self.read = read
+
+        if write is None:
+            self.write = list()  # type: List[str]
+        else:
+            self.write = write
+
+        self.use_source_key = use_source_key
+
+    def validate(self):
+        if self.interface != "content":
+            raise SlotValidationError(message="interface={}".format(self.interface))
+
+        if not self.read and not self.write:
+            raise SlotValidationError(message="`read` or `write` is required for slot")
+
+    def get_content_dirs(self, installed_path: str) -> Set[str]:
+        content_dirs = set()  # type: Set[str]
+
+        for path in self.read + self.write:
+            # Strip leading "$SNAP" and "/".
+            path = re.sub(r"^\$SNAP", "", path)
+            path = re.sub(r"^/", "", path)
+            content_dirs.add(os.path.join(installed_path, path))
+
+        return content_dirs
+
+    @classmethod
+    def from_dict(cls, slot_dict, slot_name=""):
+        slot = ContentSlot(slot_name=slot_name)
+
+        # Content directories may be nested under "source",
+        # but they cannot be in both places according to snapd:
+        # https://github.com/snapcore/snapd/blob/master/interfaces/builtin/content.go#L81
+        if "source" in slot_dict:
+            source_data = slot_dict["source"]
+            slot.use_source_key = True
+        else:
+            source_data = slot_dict
+            slot.use_source_key = False
+
+        if "read" in source_data:
+            slot.read = source_data["read"]
+
+        if "write" in source_data:
+            slot.write = source_data["write"]
+
+        return slot
+
+    def to_yaml(self):
+        props = list()
+
+        # Required keys.
+        props.append(("interface", self.interface))
+
+        if self.use_source_key:
+            source = dict()
+            if self.read:
+                source["read"] = self.read
+            if self.write:
+                source["write"] = self.write
+            props.append(("source", source))
+        else:
+            if self.read:
+                props.append(("read", self.read))
+            if self.write:
+                props.append(("write", self.write))
+
+        return OrderedDict(props)
+
+
+class DbusSlot(Slot):
+    """Representation of a snap dbus slot."""
+
+    def __init__(self, *, slot_name: str = "", bus: str = "", name: str = "") -> None:
+        super().__init__(slot_name=slot_name)
+
+        self._plug_dict = dict()  # type: Dict[str, Any]
+        self.interface = "dbus"  # type: str
```

Comment:
I might pass this to parent and make it a read only property there.

Reply:
agreed